File: php/tests/Conformance/SnapshotCasesTest.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Conformance;

use DateTimeImmutable;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;
use SimpleOrm\Dialect\SqliteDialect;
use SimpleOrm\Metadata\EntityMapLoader;
use SimpleOrm\Migrations\SchemaSnapshot;
use SimpleOrm\Tests\Support\ConformancePaths;

/**
 * The snapshot-case runner (§9, ADR-0017), mirroring
 * `dotnet/tests/SimpleOrm.Tests/ConformanceSnapshotTests.cs`: each
 * `conformance/snapshot-cases/*.json` names a fixture entity, a version, and a
 * pinned generation time, and expects the exact snapshot document — every
 * implementation must export it identically from its own native entity
 * definition. Tables export by columns; views by normalized DDL.
 */
final class SnapshotCasesTest extends TestCase
{
    /** @return list<array{0: string}> */
    public static function caseFiles(): array
    {
        return array_map(
            static fn (string $file): array => [$file],
            ConformancePaths::cases('snapshot-cases'),
        );
    }

    #[Test]
    #[DataProvider('caseFiles')]
    public function export_matches_expected_document(string $fileName): void
    {
        $path = ConformancePaths::dir('snapshot-cases') . DIRECTORY_SEPARATOR . $fileName;
        $spec = json_decode(file_get_contents($path), true, flags: JSON_THROW_ON_ERROR);

        $entityName = $spec['entity'];
        $entityClass = self::findFixtureClass($entityName);
        $map = (new EntityMapLoader())->load($entityClass);
        $dialect = new SqliteDialect();
        $asOfVersion = (int) $spec['asOfVersion'];
        $generatedAt = self::parseGeneratedAt($spec['generatedAt']);

        $kind = $spec['kind'] ?? 'table';
        $produced = $kind !== 'table'
            ? SchemaSnapshot::exportDdl($map->relationName, $kind, $dialect->createViewSql($map), $asOfVersion, $generatedAt)
            : SchemaSnapshot::export(SchemaSnapshot::fromMap($map, $dialect), $asOfVersion, $generatedAt);

        self::assertSame(
            $spec['expect'],
            json_decode($produced, true, flags: JSON_THROW_ON_ERROR),
            "produced snapshot differs from the expected document:\n{$produced}",
        );
    }

    /** @return class-string */
    private static function findFixtureClass(string $entityName): string
    {
        $class = 'SimpleOrm\\Tests\\Sample\\Models\\' . $entityName;
        if (!class_exists($class)) {
            self::fail("fixture entity {$entityName} not found under SimpleOrm\\Tests\\Sample\\Models");
        }

        return $class;
    }

    /** The C# `"o"` format carries 7 fractional digits; PHP parses at most 6, so the 7th is trimmed before parsing. */
    private static function parseGeneratedAt(string $value): DateTimeImmutable
    {
        $trimmed = preg_replace('/(\.\d{6})\d*Z$/', '$1Z', $value);

        return new DateTimeImmutable($trimmed ?? $value);
    }
}

File: go/orm/internal/conformance/snapshot_cases_test.go
package conformance

import (
	"encoding/json"
	"reflect"
	"testing"
	"time"

	"github.com/kosmassp/SimpleOrm/go/orm/internal/metadata"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/testsupport"
	"github.com/kosmassp/SimpleOrm/go/orm/migrations"
	"github.com/kosmassp/SimpleOrm/go/orm/sqlite"
)

// snapshotCase is conformance/snapshot-cases/*.json (conformance/README.md):
// a fixture entity, version, and pinned generation time in, the exact
// exported snapshot document out (mirrors ConformanceSnapshotTests.cs).
type snapshotCase struct {
	Name        string          `json:"name"`
	Entity      string          `json:"entity"`
	Kind        string          `json:"kind"`
	AsOfVersion int64           `json:"asOfVersion"`
	GeneratedAt string          `json:"generatedAt"`
	Expect      json.RawMessage `json:"expect"`
}

func TestSnapshotCases_MatchExpectedOutput(t *testing.T) {
	for _, file := range testsupport.ConformanceCases(t, "snapshot-cases") {
		t.Run(file, func(t *testing.T) {
			data := testsupport.ReadConformance(t, "snapshot-cases", file)
			var c snapshotCase
			if err := json.Unmarshal(data, &c); err != nil {
				t.Fatalf("parse case: %v", err)
			}

			entityType, ok := entityTypeNamed(c.Entity)
			if !ok {
				t.Fatalf("no fixture entity named %q", c.Entity)
			}
			loader := metadata.NewLoader(nil)
			m, err := loader.Load(entityType)
			if err != nil {
				t.Fatalf("load %s: %v", c.Entity, err)
			}
			dialect := sqlite.New()
			generatedAt, err := time.Parse(time.RFC3339Nano, c.GeneratedAt)
			if err != nil {
				t.Fatalf("parse generatedAt: %v", err)
			}

			var produced string
			if c.Kind != "" && c.Kind != "table" {
				produced = migrations.ExportDDL(m.RelationName, c.Kind, dialect.CreateViewSQL(m), c.AsOfVersion, generatedAt)
			} else {
				produced = migrations.Export(m, dialect, c.AsOfVersion, generatedAt)
			}

			var expected, actual any
			if err := json.Unmarshal(c.Expect, &expected); err != nil {
				t.Fatalf("parse expect: %v", err)
			}
			if err := json.Unmarshal([]byte(produced), &actual); err != nil {
				t.Fatalf("parse produced: %v\n%s", err, produced)
			}
			if !reflect.DeepEqual(expected, actual) {
				t.Errorf("produced snapshot differs from the expected document:\n%s", produced)
			}
		})
	}
}

// entityTypeNamed resolves a fixture entity by its unqualified type name (the
// case format's "entity"), the way the C# reference resolves it by scanning
// the sample assembly's exported types.
func entityTypeNamed(name string) (reflect.Type, bool) {
	for _, t := range entityTypes {
		if t.Name() == name {
			return t, true
		}
	}
	return nil, false
}
